consumer/consumererror/README.md

```diff
@@ -0,0 +1,113 @@
+# Consumer errors
+
+This package contains error types that should be returned by a consumer when an
+error occurs while processing telemetry. The error types included in this
+package provide functionality for communicating details about
+the error for use upstream in the pipeline. Ideally the error returned by a
+component in its `consume` function should be from this package.
+
+## Use cases
+
+**Retry logic**: Errors should be allowed to embed rejected telemetry items to
+be retried by the consumer.
+
+**Indicating partial success**: Errors can indicate that not all items were
+accepted, for example as in an OTLP partial success message.
+
+**Communicating network error codes**: Errors should allow embedding information
+necessary for the Collector to act as a proxy for a backend, i.e. relay a status
+code returned from a backend in a response to a system upstream from the
+Collector.
+
+## Creating Errors
+
+Errors can be created through one of the three constructors provided:
+
+- `consumererror.New[Signal]` for retryable errors.
+- `consumererror.NewPartial` for errors where only some of the items failed.
+- `consumererror.NewHTTPStatus` for errors resulting from an HTTP call with an error status code.
+- `consumererror.NewGRPCStatus` for errors resulting from a gRPC call with an error status code.
+
+Only retryable errors will be retried, all other errors are considered permanent
+and will not be retried.
+
+Errors can be joined by passing them to a call to `errors.Join`.
+
+
+### Retryable errors
```

Comment:
I realized we haven't documented what to do if there is a discrepancy between the data held by the caller and the consumer. The only behavior that I think will be consistent would be to only consider the data returned by the consumer. Otherwise we should remove this parameter from the `New[Signal]` functions and just use the copy held by the caller. If anyone has other thoughts on this, please let me know.

Reply:
I think these two approaches would be our best bet:

1. Continue to require data in `New[Signal]` and indicate to caller implementors that they can discard their copy of the data and rely on the consumer to return a copy. This requires consumers to keep track of the data they submitted.
2. Add a `WithRetryData` option or similar that indicates exactly which data to retry, allowing consumers to specify a subset of data if they want, and remove signal data as a required parameter in the constructor. Callers are expected to keep a copy of the data submitted to each consumer, but if they receive this option are expected to use the consumer's copy instead. I think this is more in line with how things currently work.

Reply:
Per an offline discussion on 2024-04-17, it was agreed to take approach 1.